metatensor-torch/src/atomistic.cpp

```diff
@@ -0,0 +1,986 @@
+    auto result = std::vector<NeighborsListOptions>();
+    for (const auto& it: neighbors_) {
+        result.emplace_back(it.first);
+    }
+    return result;
+}
+
+bool is_alpha_or_digit(char c) {
+    return (('0' <= c) && (c <= '9')) ||
+           (('a' <= c) && (c <= 'z')) ||
+           (('A' <= c) && (c <= 'Z'));
+}
+
+static bool valid_ident(const std::string& string) {
+    if (string.empty()) {
+        return false;
+    }
+
+    for (const auto c: string) {
+        if (!(is_alpha_or_digit(c) || c == '_' || c == '-')) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static std::string string_lower(const std::string& value) {
+    auto copy = value;
+    std::transform(copy.begin(), copy.end(), copy.begin(),
+        [](unsigned char c){ return std::tolower(c); }
+    );
+    return copy;
+}
+
+static auto INVALID_DATA_NAMES = std::unordered_set<std::string>{
+    "species",
+    "positions", "position",
+    "cell",
+    "neighbors", "neighbor"
+};
+
+void SystemHolder::add_data(std::string name, TorchTensorBlock values) {
+    if (!valid_ident(name)) {
+        C10_THROW_ERROR(ValueError,
+            "custom data name '" + name + "' is invalid: only [a-z A-Z 0-9 _-] are accepted"
+        );
+    }
+
+    if (INVALID_DATA_NAMES.find(string_lower(name)) != INVALID_DATA_NAMES.end()) {
+        C10_THROW_ERROR(ValueError,
+            "custom data can not be named '" + name + "'"
+        );
+    }
+
+    if (data_.find(name) != data_.end()) {
+        C10_THROW_ERROR(ValueError,
+            "custom data '" + name + "' is already present in this system"
+        );
+    }
+
+    const auto& values_tensor = values->values();
+    if (values_tensor.device() != this->device()) {
+        C10_THROW_ERROR(ValueError,
+            "device (" + values_tensor.device().str() + ") of the custom data "
+            "'" + name + "' does not match this system device (" + this->device().str() +")"
+        );
+    }
+
+    if (values_tensor.scalar_type() != this->scalar_type()) {
+        C10_THROW_ERROR(ValueError,
+            "dtype (" + scalar_type_name(values_tensor.scalar_type()) + ") of " +
+            "custom data '" + name + "' does not match this system " +
+            "dtype (" + scalar_type_name(this->scalar_type()) +")"
+        );
+    }
+
+    data_.emplace(std::move(name), std::move(values));
+}
+
+TorchTensorBlock SystemHolder::get_data(std::string name) const {
+    if (INVALID_DATA_NAMES.find(string_lower(name)) != INVALID_DATA_NAMES.end()) {
+        C10_THROW_ERROR(ValueError,
+            "custom data can not be named '" + name + "'"
+        );
+    }
+
+    auto it = data_.find(name);
+    if (it == data_.end()) {
+        C10_THROW_ERROR(ValueError,
+            "no data for '" + name + "' found in this system"
+        );
+    }
+
+    TORCH_WARN_ONCE(
+        "custom data '", name, "' is experimental, please contact the developers ",
```

Comment:
I would say developers of metatensor here, maybe random people will think this is a torch error and they will contact the torch developers.
Unrelated but https://github.com/google-research/bert/issues/1396

Reply:
fair point!